### ISSUE TITLE
HTTP client/connection pooling support

### ISSUE THREAD

**Issue by jterhune** [Azure/azure-storage-java#212]
We see that the current code uses Java's HttpURLConnection. It would be nice to support additional HTTP clients, which could enable connection pooling and improve memory usage and scalability.

The minimal first step would probably be creating an interface similar to Spring's [ClientHttpRequestFactory](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/client/ClientHttpRequestFactory.html) and related Request/Response interfaces, and then moving the HttpURLConnection code to be the default implementation. Then applications would be free to provide their own implementations using Feign, OkHttp, or the Apache HTTP client.

**Comment by mirobers** [comment]
This would be a pretty hefty refactor. We don't plan to do this now but we'll track it as an enhancement.

**Comment by maltalex** [comment]
I strongly support this request. 

It's not just a "nice to have" enhancement. The absence of connection pooling is a **major** performance bottleneck. In our internal benchmarking, for both the Azure Table and Azure Queue we couldn't get anywhere near their performance targets, even on a 32 core machine. The only way to get near the performance targets was to use a third party asynchrnonous http client with connection and thread pooling.

The creation of a new socket for every operation is **very** expensive and to my knowledge isn't done in the C# SDK, or in any other high performance API client of any service I can think of. Connection and thread pooling is such a standard technique, it's truly surprising not seeing it here.

To emphasize how critical the performance hit is, I suggest a back-of-the-envelope calculation: On Linux (and I assume, similarly, on Windows) the port range for outbound connections is `net.ipv4.ip_local_port_range = 32768    61000`. And the default TCP FIN timout is 60 seconds: `net.ipv4.tcp_fin_timeout = 60`. That means that a single machine cannot create more than 28,232 connections to the same destination (such as a Table/Queue endpoint) every 60 seconds before **running out of port numbers (!)** (which is of course something we've discovered in production when using the Java storage SDK...). What this means is that with these default settings (which are quite sane), in the best-case scenario, it's impossible to call the Table/Queue API more than 470.5 times/second from a single machine (<1/4 of the performance target 2K ops/sec), not to mention the toll this takes on the CPU.

In contrast, using a [Java async http client with connection pooling](https://github.com/AsyncHttpClient/async-http-client/) we quite easily managed to saturate the 10Gbps network with Table responses and get a x10 performance improvement over everything we managed to get out of the official SDK on the same machine.

**Comment by mirobers** [comment]
@maltalex You are correct that connection reuse is important. In fact, the storage client today does reuse connections, so we happily avoid the case where each request creates a new socket. You can observe this even in a network trace of a simple application that makes a few requests in series. Each request will reuse the same socket. Java's connection pooling isn't perfect, however, and I understand the desire for improvement there.

You also bring up the issue of asynchronous I/O and thread pooling. We are working on providing a library with full async support. This will enter preview in the coming months. (There's no firm date at this point.)

**Comment by maltalex** [comment]
@mirobers Thank you for your response. 

We've had issues with processes being unable to open new sockets to a queue/table endpoint once they were making a few hundred thousand requests per minute and running out of sockets. We had to implement our own storage client for these processes using the async http client I mentioned earlier.

Could you tell me when was connection reuse implemented? Perhaps we were using an old version. Also, is it possible to control the number of connections in the pool?

**Comment by mirobers** [comment]
@maltalex Connection reuse is supported by the Java framework via HttpURLConnection. There isn't much visibility into the pool unfortunately. You can use `http.maxConnections` to control the number of concurrent connections allowed to an endpoint. Maybe the issue you encountered was caused by this limit being too low.

It's also possible that you were hitting a bug in the client that was leaving dangling connections, such as [this bug fixed in version 5.3](https://github.com/Azure/azure-storage-java/blob/9556398f731ba3d373d5c455ef5a2f1cbe9b1887/ChangeLog.txt#L28).

There's more info in Oracle's documentation for [Persistent Connections](https://docs.oracle.com/javase/7/docs/technotes/guides/net/http-keepalive.html).

**Comment by maltalex** [comment]
I wasn't aware of connection reuse in `HttpURLConnection`, so yeah, we're probably using the defaults. Thank you for the tip.

Is it possible to document such gotchas in relevant performance-related documentation regarding azure storage? This has significant performance implications. Docs such as the [Azure storage performance checklist ](https://docs.microsoft.com/en-us/azure/storage/common/storage-performance-checklist) don't mention anything about connection reuse in Java.
